balancer_conn_wrappers.go

```diff
@@ -160,35 +195,133 @@ func (ccb *ccBalancerWrapper) handleSubConnStateChange(sc balancer.SubConn, s co
-// updateClientConnState forwards the clientConn update to the wrapped balancer
-// synchronously.
-//
-// Other calls from the channel like exitIdle() and handleSubConnStateChange()
-// are handled asynchronously by pushing the update onto a channel, which is
-// picked up by the watcher() goroutine and forwarded to the wrapped balancer.
-// That approach cannot be taken here because the corresponding API on the
-// balancer returns an error which needs to be sent back to the channel to be
-// forward to the resolver.
-func (ccb *ccBalancerWrapper) updateClientConnState(ccs *balancer.ClientConnState) error {
+// handleSubConnStateChange handles a SubConnState update from the update
+// channel and invokes the appropriate method on the underlying balancer.
+func (ccb *ccBalancerWrapper) handleSubConnStateChange(update *scStateUpdate) {
 	ccb.balancerMu.Lock()
-	defer ccb.balancerMu.Unlock()
-	return ccb.balancer.UpdateClientConnState(*ccs)
+	ccb.balancer.UpdateSubConnState(update.sc, balancer.SubConnState{ConnectivityState: update.state, ConnectionError: update.err})
+	ccb.balancerMu.Unlock()
+}
+
+func (ccb *ccBalancerWrapper) exitIdle() {
+	if ccb == nil || ccb.cc.GetState() != connectivity.Idle {
+		return
+	}
+	ccb.updateCh.Put(watcherUpdate{typ: updateTypeExitIdle})
+}
+
+func (ccb *ccBalancerWrapper) handleExitIdle() {
+	ccb.balancerMu.Lock()
+	ccb.balancer.ExitIdle()
+	ccb.balancerMu.Unlock()
 }
 
 func (ccb *ccBalancerWrapper) resolverError(err error) {
 	if ccb == nil {
 		return
 	}
+	ccb.updateCh.Put(watcherUpdate{typ: updateTypeResolverError, update: err})
+}
+
+func (ccb *ccBalancerWrapper) handleResolverError(err error) {
 	ccb.balancerMu.Lock()
-	defer ccb.balancerMu.Unlock()
 	ccb.balancer.ResolverError(err)
+	ccb.balancerMu.Unlock()
+}
+
+// switchTo is invoked by grpc to instruct the balancer wrapper to switch to the
+// LB policy identified by name.
+//
+// ClientConn calls newCCBalancerWrapper() at creation time. Upon receipt of the
+// first good update from the name resolver, it determines the LB policy to use
+// and invokes the switchTo() method. Upon receipt of every subsequent update
+// from the name resolver, it invokes this method.
+//
+// the ccBalancerWrapper keeps track of the current LB policy name, and skips
+// the graceful balancer switching process if the name does not change.
+func (ccb *ccBalancerWrapper) switchTo(name string) {
+	if ccb == nil {
+		return
+	}
+	ccb.updateCh.Put(watcherUpdate{typ: updateTypeSwitchTo, update: name})
+}
+
+// handleSwitchTo handles a balancer switch update from the update channel. It
+// calls the SwitchTo() method on the gracefulswitch.Balancer with a
+// balancer.Builder corresponding to name. If no balancer.Builder is registered
+// for the given name, it uses the default LB policy which is "pick_first".
+func (ccb *ccBalancerWrapper) handleSwitchTo(name string) {
+	ccb.balancerMu.Lock()
+	defer ccb.balancerMu.Unlock()
+
+	if strings.EqualFold(ccb.curBalancerName, name) {
```

Comment:
I'm not sure, but we may have decided it was okay to change our balancer registry / etc to be case-sensitive like Java and C both are already.  I'll try to figure out what decisions were made on this.  If we did intend to change, please add a TODO here to make this a case sensitive comparison.